### PR TITLE
Add edge when calculating for_each orphans

### LINKED ISSUE #22407: Referencing resource instances created by "for_each"
### Terraform Version
```
Terraform v0.12.6
+ provider.azuread v0.5.1
+ provider.azurerm v1.32.1
+ provider.helm v0.10.1
+ provider.kubernetes v1.8.1
+ provider.null v2.1.2
+ provider.random v2.1.0
```

### Terraform Configuration Files

**azure.tf**
```hcl
variable "customers" {
  type = map(object({name=string}))
}

resource "azurerm_storage_account" "provisioning-datalake" {

  for_each = var.customers

  name                              = "mydatalake${each.key}"
  resource_group_name               = "${azurerm_resource_group.provisioning-group.name}"
  location                          = "${azurerm_databricks_workspace.databricks.location}"
  account_kind                      = "StorageV2"
  account_tier                      = "Standard"
  account_replication_type          = "GRS"
  is_hns_enabled                    = true
  enable_advanced_threat_protection = true

  tags = {
    environment = var.environment
    customer = each.value.name
  }
}

resource "azurerm_key_vault_secret" "key-vault-datalake-secret" {

  for_each = var.customers

  name         = "mydatalake-shared-key-${each.key}"
  value        = azurerm_storage_account.provisioning-datalake[each.key].primary_access_key
  key_vault_id = azurerm_key_vault.key-vault.id

  tags = {
    environment = var.environment
    customer = each.value.name
  }
}
```

**build.tfvars**
```hcl
environment = "Build"
customers = {
  dev = {
    name = "Development"
  },
  int = {
    name = "Integration"
  },
  stg = {
    name = "Staging"
  }
}
```

### Debug Output
[plan.log](https://gist.github.com/jallaix/e4168a9d0019dedc8b21ff0448b8a3b9)

### Expected Behavior
I expect "keyvault-datalake-secret" entries to be created with the matching keys of the generated "provisioning-datalake" resources.

### Actual Behavior
When I run ``terraform plan --var-file=build.tfvars``, I get the following error:
```
Error: Invalid index

  on azure.tf line 45, in resource "azurerm_key_vault_secret" "key-vault-datalake-secret":
  45:   value        = azurerm_storage_account.provisioning-datalake[each.key].primary_access_key
    |----------------
    | azurerm_storage_account.provisioning-datalake is object with 52 attributes
    | each.key is "stg"

The given key does not identify an element in this collection value.


Error: Invalid index

  on azure.tf line 45, in resource "azurerm_key_vault_secret" "key-vault-datalake-secret":
  45:   value        = azurerm_storage_account.provisioning-datalake[each.key].primary_access_key
    |----------------
    | azurerm_storage_account.provisioning-datalake is object with 52 attributes
    | each.key is "int"

The given key does not identify an element in this collection value.


Error: Invalid index

  on azure.tf line 45, in resource "azurerm_key_vault_secret" "key-vault-datalake-secret":
  45:   value        = azurerm_storage_account.provisioning-datalake[each.key].primary_access_key
    |----------------
    | azurerm_storage_account.provisioning-datalake is object with 52 attributes
    | each.key is "dev"

The given key does not identify an element in this collection value.
```

### References

This was originally discussed [on Stack Overflow](https://stackoverflow.com/questions/57412821/referencing-resource-instances-created-by-for-each-in-terraform).

### PR DESCRIPTION
**Pull request by pselle**
In the scenario when a resource is orphaned by the addition of for_each, in the evaluation of orphans, it is possible for the EachMode of the resource to be set incorrectly, because the last evaluated resource sets the EachMode for the whole resource. As such, adding an edge when evaluating the movement between NoKey and StringKey resources will ensure order is maintained, and an instance with StringKey will "win".

Fixes #22407

### CODE REVIEW

File: terraform/transform_orphan_count.go [hunk 1]
```diff
@@ -50,20 +50,47 @@ func (t *OrphanResourceCountTransformer) Transform(g *Graph) error {
...
+	}
+
 	for key := range haveKeys {
 		s, _ := key.(addrs.StringKey)
 		// If the key is present in our current for_each, carry on
 		if _, ok := t.ForEach[string(s)]; ok {
 			continue
 		}
 
+		// If the key is no-key, we have already added it, so skip
```
Comment (jbardin) [inline terraform/transform_orphan_count.go]: I would put this check right at the start of the loop. I think it looks better at least to indicate it has no bearing on the other other `ForEach` check and we know we've already added it.

File: terraform/transform_orphan_count.go [hunk 2]
```diff
@@ -50,20 +50,47 @@ func (t *OrphanResourceCountTransformer) Transform(g *Graph) error {
 }
 
 func (t *OrphanResourceCountTransformer) transformForEach(haveKeys map[addrs.InstanceKey]struct{}, g *Graph) error {
+	// If there is a no-key node, add this to the graph first,
+	// so that we can create edges to it in subsequent (StringKey) nodes.
+	// This is because the last item determines the resource mode for the whole resource,
+	// so if this (non-deterministically) happens to end up as the last one,
+	// that will change the resource's EachMode and our addressing for our instances
+	// will not work as expected
```
Comment (jbardin) [inline terraform/transform_orphan_count.go]: In other words, we need to evaluate the orphaned resource before the in-memory state is updated to deal with the new for_each resource. I'm not sure if that makes anything more clear or not for the comment, so I'll let you decide if you want to change the wording or not.

File: terraform/transform_orphan_count.go (hunk 1 above)
Comment (pselle) [reply terraform/transform_orphan_count.go]: I agree, updating

File: terraform/transform_orphan_count.go (hunk 2 above)
Comment (pselle) [reply terraform/transform_orphan_count.go]: I've made a couple tweaks -- mind taking a look? I'm being a bit wordy to also try and communicate my own understanding, so do appreciate the read of this :)

Comment (jbardin) [reply terraform/transform_orphan_count.go]: I think the new wording works :+1:
